pennylane/plugins/tests/conftest.py

```diff
@@ -0,0 +1,136 @@
+            return float(os.environ.get("TOL", TOL))
+        else:
+            return TOL_STOCHASTIC
+
+    return _tol
+
+
+@pytest.fixture(scope="session")
+def init_state():
+    """Fixture to create an n-qubit random initial state vector."""
+
+    def _init_state(n):
+        state = np.random.random([2 ** n]) + np.random.random([2 ** n]) * 1j
+        state /= np.linalg.norm(state)
+        return state
+
+    return _init_state
+
+
+@pytest.fixture(scope="session")
+def skip_if():
+    """Fixture to skip tests."""
+
+    def _skip_if(condition):
+        if condition:
+            pytest.skip("Test does not apply to this device.")
+
+    return _skip_if
+
+
+@pytest.fixture(scope="function")
+def device(device_kwargs):
+    """Fixture to create a device."""
+
+    def _device(n_wires):
+        device_kwargs["wires"] = n_wires
+        return qml.device(**device_kwargs)
+
+    return _device
+
+
+# ============================
+# These functions are required to define the device name to run the tests for
+
+
+def pytest_addoption(parser):
+    """Add command line option to pytest."""
+
+    # TODO: find a way to pass an unknown number of optional command line options to pytest
+    # so we can optionally provide compulsory args like "cutoff_dim" or specific args such as a backend
+
+    parser.addoption("--device", action="store", default=None)
+    parser.addoption("--shots", action="store", default=None, type=int)
+    parser.addoption("--analytic", action="store", default=None)
+
+
+def pytest_generate_tests(metafunc):
+    """Set up fixtures from command line options. """
+    opt = metafunc.config.option
+
+    device_kwargs = {
+        "name": opt.device,
+        "shots": opt.shots,
+        "analytic": opt.analytic,
+    }
+
+    # =======================
+    # some processing of the command line options
+    if device_kwargs["shots"] is None:
+        device_kwargs["shots"] = N_SHOTS
+    if device_kwargs["analytic"] is None:
+        # use default value of device
+        device_kwargs.pop("analytic")
+    else:
+        # turn string into boolean
+        if device_kwargs["analytic"] == "False":
+            device_kwargs["analytic"] = False
+        else:
+            device_kwargs["analytic"] = True
+    # ====================
+
+    # parametrize function if device_kwargs is an argument
+    # this is needed for the "device" fixture
+    if "device_kwargs" in metafunc.fixturenames:
+        if opt.device is None:
+            # if no command line argument for device given, run tests on core devices
+            list_device_kwargs = []
+            for dev_name in LIST_CORE_DEVICES:
+                core_dev_kwargs = device_kwargs.copy()
+                core_dev_kwargs["name"] = dev_name
+                list_device_kwargs.append(core_dev_kwargs)
+
+            metafunc.parametrize("device_kwargs", list_device_kwargs)
+        else:
+            metafunc.parametrize("device_kwargs", [device_kwargs])

```

Comment:
Small thought - I noticed below that there are some `skip_if` conditions that are repeated a lot. We could turn these into mark decorators, like we do in SF:

```python
@pytest.mark.backends("gaussian", "fock")
class TestSomething:
```

The nice thing about marks is you can apply them to classes rather than each test, and you can even apply them to an entire file:

```python
mark = pytest.mark.backends("tf")
```

To do this in SF, we define the `pytest_runtest_setup()` function, which is executed prior to each test.

```python
def pytest_runtest_setup(item):
    """Automatically skip tests if they are marked for only certain backends"""
    if tf_available:
        allowed_backends = {"gaussian", "tf", "fock"}
    else:
        allowed_backends = {"gaussian", "fock"}

    # load the marker specifying what the backend is
    marks = {mark.name for mark in item.iter_markers() if mark.name in allowed_backends}

    # load a marker specifying whether the test only works with certain backends
    test_backends = [mark.args for mark in item.iter_markers(name="backends")]

    if not test_backends:
        # if the test hasn't specified that it runs on particular backends,
        # assume it will work with all backends
        test_backends = allowed_backends
    else:
        # otherwise, extract the set of backend strings the test works with
        test_backends = test_backends[0]

    for b in marks:
        if b not in test_backends:
            pytest.skip(
                "\nTest {} only runs with {} backend(s), "
                "but {} backend provided".format(item.nodeid, test_backends, b)
            )

    # skip broken tests
    for mark in item.iter_markers(name="broken"):
        if mark.args:
            pytest.skip("Broken test skipped: {}".format(*mark.args))
        else:
            pytest.skip("Test skipped as corresponding code base is currently broken!")
```
(I don't think ours needs to look this complicated).

Reply:
This way, we could have
```python
@pytest.mark.model("qubit")
```

Reply:
That is really cool. But it is again a bit more opaque. Skip_if will tell every coder what the test is run for, while for "mark" there is magic happening...Is this a case of pytest genius versus clarity?  Or do you think it would be more readable with the `mark` decorator?

Reply:
Oh, I see `mark` used quite often in other Python packages, so it feels relatively 'normal' to me.

Reply:
Instead of programmatically separating qubit and cv cases, could we create separate folders for each set of test files (e.g. `tests/qubit` & `tests/cv`)? And files added in this PR would mainly go into `tests/qubit`.

It would feel natural to me that we have the branching logic earlier in the line of execution. This would also help with the organization of the files. (Still picking up details of how the cli for pytest really propagates the info on which test cases to call, so let me know if this would be difficult to achieve)

I'd imagine that actually creating two separate test suites might also not be an entirely radical idea. There would be code duplication involved, but I'd mostly think we are either working on the qubit based parts or the CV based parts. Exceptions could include extending e.g. shared parts of nodes, but we'd have to add different integration tests anyways.

Reply:
I'm in favour of that @antalszava, since CV/Qubit only share very few tests anyways, and may require different base settings for the stochastic tests...